plugins/tickets.py

```diff
@@ -0,0 +1,1821 @@
+);
+
+CREATE FUNCTION tickets.log_ticket_update() RETURNS TRIGGER AS $log_ticket_update$
+        DECLARE
+                modified tickets.tickets%rowtype;
+                last_version int;
+        BEGIN
+                SELECT INTO modified
+                        NEW.id,
+                        NULLIF(NEW.type, OLD.type),
+                        NULLIF(NEW.stage, OLD.stage),
+                        NULLIF(NEW.status, OLD.status),
+                        NULLIF(NEW.modid, OLD.modid),
+                        NULLIF(NEW.targetid, OLD.targetid),
+                        NULLIF(NEW.roleid, OLD.roleid),
+                        NULLIF(NEW.auditid, OLD.auditid),
+                        NULLIF(NEW.duration, OLD.duration),
+                        NULLIF(NEW.comment, OLD.comment),
+                        NULLIF(NEW.list_msgid, OLD.list_msgid),
+                        NULLIF(NEW.delivered_id, OLD.delivered_id),
+                        NULLIF(NEW.created_at, OLD.created_at),
+                        NEW.modified_by;
+                SELECT version INTO last_version FROM tickets.history WHERE id = OLD.id ORDER BY version DESC LIMIT 1;
+                IF NOT FOUND THEN
+                        INSERT into tickets.history VALUES (0, OLD.created_at, OLD.*), (1, now(), modified.*);
+                ELSE
+                        INSERT into tickets.history VALUES (coalesce(last_version + 1, 1), now(), modified.*);
+                END IF;
+                RETURN NULL;
+        END
+$log_ticket_update$ LANGUAGE plpgsql;
+
+CREATE TRIGGER log_update
+        AFTER UPDATE ON tickets.tickets
+        FOR EACH ROW
+        WHEN (OLD.* IS DISTINCT FROM NEW.*)
+        EXECUTE PROCEDURE tickets.log_ticket_update();
+"""
+
+ticket_comment_re = re.compile(
+    r"""
+    (?i)\s*
+    ([\d.]+)
+    (s(?:ec(?:ond)?s?)?
+    |(?-i:m)|min(?:ute)?s?
+    |h(?:(?:ou)?rs?)?
+    |d(?:ays?)?
+    |w(?:(?:ee)?ks?)
+    |(?-i:M)|months?
+    |y(?:(?:ea)?rs?)?
+    )
+    |p(?:erm(?:anent)?)?\W+
+    """.replace(' ', '').replace('\n', '')
+)
+
+time_expansion = {
+    's': 1,
+    'm': 60,
+    'h': 60 * 60,
+    'd': 60 * 60 * 24,
+    'w': 60 * 60 * 24 * 7,
+    'M': 60 * 60 * 24 * 30,
+    'y': 60 * 60 * 24 * 365
+}
+
+# ----------- Config -----------
+conf = db.kv.Config(__name__)  # General plugin configuration
+
+conf.guild: int  # ID of the guild the ticket system is managing
+conf.tracked_roles: List[int]  # List of roleids of tracked roles
+conf.last_audit_id: int  # ID of last audit event processed
+conf.ticket_list: int  # Channel id of the ticket list in the guild
+
+
+# ----------- Data -----------
+@init_for(__name__)
+def init():
+    return ticket_schema
+
+
+class fieldConstants(Enum):
+    """
+    A collection of database field constants to use for selection conditions.
+    """
+    NULL = "IS NULL"
+    NOTNULL = "IS NOT NULL"
+
+
+class _rowInterface:
+    __slots__ = ('row', '_pending')
+
+    _conn = db.connection()
```

Comment:
one connection for the entire class and all its instances?

Reply:
Is it really that high? I don't anticipate more than 1 per second. Also connections over unix domain socket are essentially IPC.

Reply:
Though looking into pooling is still a good idea.